crates/agentgateway/src/control/mod.rs

```diff
@@ -71,46 +72,135 @@ pub enum AuthSource {
 	None,
 }
 
-impl AuthSource {
-	pub async fn insert_headers(&self, request: &mut http::HeaderMap) -> anyhow::Result<()> {
-		const AUTHORIZATION: &str = "authorization";
-		const CLUSTER: &str = "clusterid";
-		match self {
+#[derive(serde::Serialize, Clone)]
+struct AuthSourceLoaderInner {
+	cluster_id: String,
+	#[serde(serialize_with = "ser_redact")]
+	current_token: Arc<RwLock<Arc<Vec<u8>>>>,
+}
+
+impl fmt::Debug for AuthSourceLoaderInner {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		f.debug_struct("AuthSourceLoader")
+			.field("cluster_id", &self.cluster_id)
+			.field("current_token", &"<redacted>")
+			.finish()
+	}
+}
+
+#[derive(serde::Serialize, Debug)]
+pub struct AuthSourceLoader {
+	inner: Option<AuthSourceLoaderInner>,
+	#[serde(skip)]
+	drop_notifier: Option<tokio::sync::oneshot::Sender<()>>,
+}
+
+impl Drop for AuthSourceLoader {
+	fn drop(&mut self) {
+		if let Some(tx) = self.drop_notifier.take() {
+			let _ = tx.send(());
+		}
+	}
```

Comment:
I think just dropping it is actually good enough (so no need to impl Drop at all)